File: translate_qa.py
import re
from bs4 import BeautifulSoup
from deep_translator import GoogleTranslator
import time

def translate_to_hindi(text):
    text = text.strip()
    if not text:
        return text
    try:
        translated = GoogleTranslator(source='en', target='hi').translate(text)
        time.sleep(0.1) # Be nice to the API
        return translated
    except Exception as e:
        print(f"Error translating: {e}")
        return text

with open('index.html', 'r', encoding='utf-8') as f:
    html = f.read()

soup = BeautifulSoup(html, 'html.parser')

accordions = soup.find_all('details', class_='qa-accordion')
print(f"Found {len(accordions)} accordions to translate.")

count = 0
for acc in accordions:
    # 1. Translate Summary text
    summary = acc.find('summary')
    if summary:
        # Check if already translated (contains .en or .hi)
        if not summary.find('span', class_='en'):
            # It usually has a badge <span class="qa-badge">Q1</span> and then text
            badge = summary.find('span', class_='qa-badge')
            marks = summary.find('span', class_='qa-marks')
            
            # We want to extract the text, excluding badge and marks
            # To do this safely, let's reconstruct the summary
            if badge:
                badge_html = str(badge)
                badge.extract()
            else:
                badge_html = ""
                
            if marks:
                marks_html = str(marks)
                marks.extract()
            else:
                marks_html = ""
                
            original_text = summary.get_text(strip=True)
            if original_text:
                hindi_text = translate_to_hindi(original_text)
                new_html = f"{badge_html} <span class=\"hi\">{hindi_text}</span><span class=\"en\">{original_text}</span> {marks_html}"
                # Replace summary contents
                summary.clear()
                summary.append(BeautifulSoup(new_html, 'html.parser'))
                count += 1
                
    # 2. Translate QA Content
    content = acc.find('div', class_='qa-content')
    if content:
        # Find all p, li, h5 that do not contain .hi or .en
        for tag_name in ['p', 'li', 'h5']:
            for tag in content.find_all(tag_name):
                if not tag.find('span', class_='en'):
                    # To handle strong tags inside p or li, we will just translate the entire HTML content of the tag
                    # Wait, deep-translator works best with raw text. 
                    # If we translate text and lose HTML (like strong), that's bad.
                    # A safe way is to just translate the raw text, but maybe it breaks formatting.
                    # Since Google Translate API can handle basic HTML sometimes, let's try passing the inner HTML? No, deep-translator strips HTML or gets confused.
                    
                    # Instead of translating the whole inner HTML, let's just translate the text.
                    # For simplicity, since it's just Q&A, we can translate the get_text().
                    original_html = "".join(str(item) for item in tag.contents)
                    original_text = tag.get_text(strip=True)
                    if original_text:
                        hindi_text = translate_to_hindi(original_text)
                        
                        # Just wrap the original HTML in .en and translated text in .hi
                        new_html = f"<span class=\"hi\">{hindi_text}</span><span class=\"en\">{original_html}</span>"
                        tag.clear()
                        tag.append(BeautifulSoup(new_html, 'html.parser'))
                        count += 1

    if count % 10 == 0:
        print(f"Processed {count} elements...")

with open('index.html', 'w', encoding='utf-8') as f:
    f.write(str(soup))

print(f"Done. Translated {count} elements.")

File: translate_qa.js
const fs = require('fs');
const jsdom = require('jsdom');
const { JSDOM } = jsdom;

// The @vitalets/google-translate-api works via fetch
async function run() {
    const { translate } = await import('@vitalets/google-translate-api');
    
    let html = fs.readFileSync('index.html', 'utf-8');
    const dom = new JSDOM(html);
    const document = dom.window.document;

    const accordions = document.querySelectorAll('details.qa-accordion');
    console.log(`Found ${accordions.length} accordions.`);

    let promises = [];
    let count = 0;

    for (let acc of accordions) {
        // Translate Summary
        let summary = acc.querySelector('summary');
        if (summary && !summary.querySelector('.en')) {
            let badge = summary.querySelector('.qa-badge');
            let marks = summary.querySelector('.qa-marks');
            
            let badgeHTML = badge ? badge.outerHTML : '';
            let marksHTML = marks ? marks.outerHTML : '';
            
            if (badge) badge.remove();
            if (marks) marks.remove();
            
            let originalText = summary.textContent.trim();
            if (originalText) {
                promises.push((async () => {
                    try {
                        let res = await translate(originalText, { to: 'hi' });
                        let newHTML = `${badgeHTML} <span class="hi">${res.text}</span><span class="en">${originalText}</span> ${marksHTML}`;
                        summary.innerHTML = newHTML;
                        count++;
                    } catch (e) {
                        console.error('Error translating summary:', e);
                    }
                })());
            }
        }

        // Translate content
        let content = acc.querySelector('.qa-content');
        if (content) {
            let elements = content.querySelectorAll('p, li, h5');
            for (let el of elements) {
                if (!el.querySelector('.en')) {
                    let originalHTML = el.innerHTML;
                    let originalText = el.textContent.trim();
                    if (originalText) {
                        promises.push((async () => {
                            try {
                                let res = await translate(originalText, { to: 'hi' });
                                let newHTML = `<span class="hi">${res.text}</span><span class="en">${originalHTML}</span>`;
                                el.innerHTML = newHTML;
                                count++;
                            } catch (e) {
                                console.error('Error translating element:', e);
                            }
                        })());
                    }
                }
            }
        }
    }

    console.log(`Waiting for ${promises.length} translation requests...`);
    
    // Process in batches of 10 to avoid rate limits
    const batchSize = 10;
    for (let i = 0; i < promises.length; i += batchSize) {
        const batch = promises.slice(i, i + batchSize);
        await Promise.all(batch);
        console.log(`Completed batch ${i / batchSize + 1} of ${Math.ceil(promises.length / batchSize)}`);
        await new Promise(r => setTimeout(r, 500)); // 500ms delay between batches
    }

    // Write back
    fs.writeFileSync('index.html', dom.serialize(), 'utf-8');
    console.log(`Done. Translated elements.`);
}

run().catch(console.error);
